src/S3/S3Transfer/TransferListener.php

```diff
@@ -0,0 +1,292 @@
+<?php
+
+namespace Aws\S3\S3Transfer;
+
+use Closure;
+use Throwable;
+
+class TransferListener extends ListenerNotifier
+{
+    /**
+     * @param Closure|null $onTransferInitiated
+     * No parameters will be passed.
+     *
+     * @param Closure|null $onObjectTransferInitiated
+     *  Parameters that will be passed when invoked:
+     *  - $objectKey: The key that identifies the object being transferred.
+     *  - $objectRequestArgs: The arguments that initiated the object transfer request.
+     *
+     * @param Closure|null $onObjectTransferProgress
+     *  Parameters that will be passed when invoked:
+     *  - $objectKey: The key that identifies the object being transferred.
+     *  - $objectBytesTransferred: The total of bytes transferred for this object.
+     *  - $objectSizeInBytes: The size in bytes of the object.
+     *
+     * @param Closure|null $onObjectTransferFailed
+     *  Parameters that will be passed when invoked:
+     *  - $objectKey: The object key for which the transfer has failed.
+     *  - $objectBytesTransferred: The total of bytes transferred from
+     *    this object.
+     *  - $reason: The reason why the transfer failed for this object.
+     *
+     * @param Closure|null $onObjectTransferCompleted
+     * Parameters that will be passed when invoked:
+     *  - $objectKey: The object key for which the transfer was completed.
+     *  - $objectBytesCompleted: The total of bytes transferred for this object.
+     *
+     * @param Closure|null $onTransferProgress
+     * Parameters that will be passed when invoked:
+     *  - $totalObjectsTransferred: The number of objects transferred.
+     *  - $totalBytesTransferred: The total of bytes already transferred on this event.
+     *  - $totalBytes: The total of bytes to be transferred.
+     *
+     * @param Closure|null $onTransferCompleted
+     * Parameters that will be passed when invoked:
+     *  - $objectsTransferCompleted: The number of objects that were transferred.
+     *  - $objectsBytesTransferred: The total of bytes that were transferred.
+     *
+     * @param Closure|null $onTransferFailed
+     * Parameters that will be passed when invoked:
+     *  - $objectsTransferCompleted: The total of objects transferred before failure.
+     *  - $objectsBytesTransferred: The total of bytes transferred before failure.
+     *  - $objectsTransferFailed: The total of objects that failed in the transfer.
+     *  - $reason: The throwable with the reason why the transfer failed.
+     * @param int $objectsTransferCompleted
+     * @param int $objectsBytesTransferred
+     * @param int $objectsTransferFailed
+     * @param int $objectsToBeTransferred
+     */
+    public function __construct(
```

Comment:
Could all of these callbacks be consolidated into a single array or config object?

Reply:
This implementation has been replaced. Now, there is an abstract class that customers can implement and override the methods they want to listen to.